### DISCUSSION TITLE
Deployment method - Working changes

### DISCUSSION THREAD

**Discussion by dennisnel** [codefori/vscode-ibmi#700] [Q&A] [answered]
Hi there,

Could someone please explain the "new" deployment method **Working Changes**?

What is the definition of **Working Changes**?

Tx

**Comment by worksofliam** [comment, marked as answer]
This is related to git. Selecting 'working changes' will only upload files that git sees as changed that are not staged.

**Reply by dennisnel** [reply]
Thanks,

So the "workflow", when working in your git branch, is ;
- **"Working Changes"** before stashing and committing
- **"Stashed"** the changes that you have _Stashed_
- **ALL** is obvious

Is there a way that once we have _Commited_ to the branch, that we can identify only the source members that have changed? And then add a deployment method for that?

The reason being that I would only want to "deploy" the changes in the branch rather than ALL the source members.

**Reply by andreas-prouza** [reply]
I don't use this feature, I always use RSYNC to synchronise all sources from my local machine to my IBM i working directory.
I don't care if the source is staged, commited or not.
This is specially helpful if you work with different branches in GIT.

I use gnu make on IBM i to build the sources.
In gnu make (gmake) you can run the build process but without to run the final compile command for each object.
`gmake all -n`
-n means "dry run"
With this you can see which objects would be compiled.
I set up my makefile (gnu make configuration) to print out some information for each compile to the console.
So I am able to just filter (grep) for that information.
With that I can say: 

- print a list with only the object names which would be compiled
- print a list with object names and compile command for these changed sources

In general BOB (better object builder) is doing the same: rsync + gmake

**Reply by worksofliam** [reply]
@andreas-prouza by any chance do you have a task setup in VS Code to run rsync for you? Or are you running it manually?

**Reply by andreas-prouza** [reply]
Hi Liam,
I use 2 extensions for that:

- Run on Save: 
   Automatically sync when source is saved. 
   So no need to do some extra sync
- Command Runner
  To trigger the sync manually (e.g. if I switch branch) using STRG+P --> Select: Run Command --> Select: Sync project

```.vscode/settings.json```
```json
{
    "emeraldwalk.runonsave": {
        "commands": [
            {
                "match": ".*/myapp.*/.*$",
                "isAsync": true,
                "cmd": "rsync -av --rsync-path=/QOpenSys/pkgs/bin/rsync --exclude={'.git','build','logs','.vscode','.project','.gitignore'} /home/andreas/projects/myapp/ ibmi:/home/prouza/myapp/"
            }
        ]
    },
    "command-runner.terminal.name": "Sync project",
    "command-runner.terminal.autoClear": true,
    "command-runner.terminal.autoFocus": true,
    "command-runner.commands": {
        "Sync source to IBM i": "rsync -av --rsync-path=/QOpenSys/pkgs/bin/rsync --exclude={'.git','build','logs','.vscode','.project','.gitignore'} /home/andreas/projects/myapp/ ibmi:/home/prouza/myapp/",
        "Sync logs back": "rsync -avz --rsync-path=/QOpenSys/pkgs/bin/rsync --include={'logs/***','build/***'} --exclude='*' ibmi:/home/prouza/myapp/ /home/andreas/projekte/projects/myapp/",
    }
}
```
Just for info: I am developing with a Linux machine. So my local directory is ```/home/andreas``` instead of C:\users\andreas if you working on a Windows machine.
On Windows you also need to install rsync e.g. via Cygwin.
In Windows you need also to consider the backslash instead of slash in your ```settings.json```.

**Reply by andreas-prouza** [reply]
In addition I also use the "Work with Actions" possibility in your extension to build the application with gmake.
In "Command to run" I use:
```sh
/QOpenSys/pkgs/bin/bash -c "error=*EVENTF lib1=&CURLIB cd ~/myapp; gmake all"
```
The advantage of this is, that I get the compile information directly in my currently opened source.
So if I am trouble shooting with a source change, this is my choise.
As far as I know it even works if my opened source is a local one and not opened from IBM i IFS.

**Reply by andreas-prouza** [reply]
I also use the "Work with Actions" to show a list of changed sources which would be compiled (what I already described in my first answer).
This is also greate to do it in vscode directly and no need to use a bash terminal for it.

**Reply by worksofliam** [reply]
@andreas-prouza Very cool - thanks!

Just so you know, you can also create a `.vscode/actions.json` file which can then keep your compile options as part of your repo.

```json
[{
    "command": "/QOpenSys/pkgs/bin/bash -c \"error=*EVENTF lib1=&CURLIB cd ~/myapp; gmake all\"",
    "deployFirst": false, // rsync handles this!
    "environment": "pase",
    "extensions": [
        "GLOBAL"
    ],
    "name": "Run build"
},
{
    "command": "/QOpenSys/pkgs/bin/bash -c \"error=*EVENTF lib1=&CURLIB cd ~/myapp; gmake all -n\"",
    "deployFirst": false,
    "environment": "pase",
    "extensions": [
        "GLOBAL"
    ],
    "name": "Dry run content"
}]
```

**Reply by andreas-prouza** [reply]
Thanks for the hint Liam!
Cuttently I don't sync the .vscode directory, but it would makes sense to add the actions.json to it.
